### PR TITLE
add `/GET` prototype endpoint

### PR DESCRIPTION
**Pull request by isabelizimm**
Need to fix up some tests, but adding the ability to look at prototype from API endpoint

### CODE REVIEW

**Comment by juliasilge** [comment]
I would like to make something that is pretty unified between R and Python for this, because this is a part of dealing with deployed models that it would be reasonable to want to do from both languages ("for model X, what is the input data prototype?"), or even just using curl or another way to query a REST API.

Given that, this seems awkwardly deeply nested to me:

```json
{
  "title": "my_model prototype",
  "$ref": "#/definitions/prototype",
  "definitions": {
    "prototype": {
      "title": "prototype",
      "type": "object",
      "properties": {
        "B": {
          "title": "B",
          "default": 6,
          "type": "integer"
        },
        "C": {
          "title": "C",
          "default": 10,
          "type": "integer"
        },
        "D": {
          "title": "D",
          "default": 8,
          "type": "integer"
        }
      }
    }
  }
}
```

My initial inclination is that something more like this would be more usable generally:

```json
{
  "title": "my_model prototype",
  "prototype": {
    "B": {
      "title": "B",
      "default": 6,
      "type": "integer"
    },
    "C": {
      "title": "C",
      "default": 10,
      "type": "integer"
    },
    "D": {
      "title": "D",
      "default": 8,
      "type": "integer"
    }
  }
}
```

How doable would it be to go from something like that to a Pydantic schema or base model?

**Comment by juliasilge** [comment]
That example looks _way_ better, and your plan to make a little helper function sounds like the right move to me. 👍

**Comment by juliasilge** [comment]
I was working on this in a new package: https://github.com/juliasilge/cereal

Here is what I have going so far:

```json
{
    "title": "my_model prototype",
    "properties": {
        "a": {
            "class": "double",
            "details": []
        },
        "b": {
            "class": "integer",
            "details": []
        },
        "c": {
            "class": "Date",
            "details": []
        },
        "d": {
            "class": "POSIXct",
            "details": {
                "tzone": "America/New_York"
            }
        },
        "e": {
            "class": "character",
            "details": []
        },
        "f": {
            "class": "factor",
            "details": {
                "levels": ["blue", "green", "red"]
            }
        },
        "g": {
            "class": "ordered",
            "details": {
                "levels": ["small", "medium", "large"]
            }
        }
    }
}
```

- I can definitely change to `"type"` instead of `"class"`.
- It may make sense for the `details` element to not exist in Python, given that there is less use of complex data types. What do you think?
- What do you think about not having the `"title"` info duplicated in every element? Is that doable?

**Comment by isabelizimm** [comment]
> * I can definitely change to `"type"` instead of `"class"`.

At least for the Python side, I do think `"type"` should keep the name.

> * It may make sense for the `details` element to not exist in Python, given that there is less use of complex data types. What do you think?

I *could* add the details if the type is categorical/timedelta/etc, but that would all likely have to be written by hand. It's doable, but likely not worth it for the first iteration.

> * What do you think about not having the `"title"` info duplicated in every element? Is that doable?

It's possible to pop `"title"` from each element! Since there will need to be a helper function, I can remove that from the typed dict.

**Comment by isabelizimm** [comment]
What would a more complex example look like? ie, BaseModel w/ only integers btwn 0-5

**Comment by juliasilge** [comment]
I've got what will be at `/prototype` from R looking like this now:

```json
{
  "title": "my_model prototype",
  "properties": {
    "a": {
      "type": "numeric",
      "default": "1",
      "details": []
    },
    "b": {
      "type": "integer",
      "default": "2",
      "details": []
    },
    "c": {
      "type": "Date",
      "default": "2023-01-01",
      "details": []
    },
    "d": {
      "type": "POSIXct",
      "default": "2019-01-01",
      "details": {
        "tzone": "America/New_York"
      }
    },
    "e": {
      "type": "character",
      "default": "x",
      "details": []
    },
    "f": {
      "type": "factor",
      "default": "blue",
      "details": {
        "levels": ["blue", "green", "red"]
      }
    },
    "g": {
      "type": "ordered",
      "default": "small",
      "details": {
        "levels": ["small", "medium", "large"]
      }
    }
  }
}
```

**Comment by juliasilge** [comment]
Do you know why the name for the example/default field ends up "default"? I was looking at the OpenAPI specification again, and [it looks like](https://swagger.io/docs/specification/adding-examples/) it would be better to call it "example".

**Comment by juliasilge** [comment]
I'm realizing as I work on the workshop materials that this has not been merged in so Python models don't have this endpoint. What still needs to be done to finish this up?

**Comment by isabelizimm** [comment]
> What still needs to be done to finish this up?

I think the helper function (if we want full parity with the R side), and then a bit of refactoring in the tests, potentially trying to round trip back into a pydantic model.

**Comment by isabelizimm** [comment]
we are still using default over example, but let's handle that in a different PR since it will include some plumbing in the prototype creation 👍

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -115,6 +115,23 @@ jobs:
         run: |
           pytest vetiver/tests/test_sklearn.py
 
+  test-pydantic-old:
```
Comment (isabelizimm) [inline .github/workflows/tests.yml:118]: adding tests for pre/post pydantic v2 until all dependencies (`spacy` does not support v2) support v2

File: vetiver/tests/test_server.py [hunk 2]
```diff
@@ -46,3 +74,32 @@ def test_get_metadata(client):
         "required_pkgs": ["scikit-learn"],
         "python_version": list(sys.version_info),  # JSON will return a list
     }
+
+
+def test_get_prototype(client):
+    response = client.get("/prototype")
+    assert response.status_code == 200, response.text
+    assert response.json() == {
```
Comment (isabelizimm) [inline vetiver/tests/test_server.py:82]: standard output for a basic prototype

File: vetiver/tests/test_server.py [hunk 3]
```diff
@@ -46,3 +74,32 @@ def test_get_metadata(client):
...
+        },
+        "title": "prototype",
+        "type": "object",
+    }
+
+
+def test_complex_prototype(complex_prototype_model):
+    response = complex_prototype_model.get("/prototype")
+    assert response.status_code == 200, response.text
+    assert response.json() == {
```
Comment (isabelizimm) [inline vetiver/tests/test_server.py:96]: more complex prototype

**Comment by isabelizimm** [comment]
After looking at the options, a more nested approach better respects prototypes made from `pydantic.BaseModel` objects since they often add extra data to the outer levels. These fields allow us the opportunity to add round trip support for very advanced prototypes in the future. @juliasilge WDYT?

**Comment by juliasilge** [comment]
Let's prioritize the experience of round-tripping a prototype for Python the way we can round-trip a prototype for R _over_ having `/prototype` endpoints that look the same in Python and R. I think that's a reasonable choice here. 👍

**Pull request merged**